redis/tests/test_cache.rs

```diff
@@ -0,0 +1,307 @@
+            .arg("key_1")
+            .query_async(&mut con)
+            .await
+            .unwrap();
+        assert_eq!(val, "1");
+        {
+            let s = con.get_cache_statistics();
+            assert_eq!(s.hit, 1);
+            assert_eq!(s.miss, 2);
+            assert_eq!(s.invalidate, 1);
+        }
+        Ok::<_, RedisError>(())
+    })
+    .unwrap();
+}
+
+#[test]
+fn test_cache_mget() {
+    use redis::ProtocolVersion;
+
+    let ctx = TestContext::new();
+    if ctx.protocol == ProtocolVersion::RESP2 {
+        return;
+    }
+
+    block_on_all(async move {
+        let mut con = ctx.multiplexed_async_connection_tokio_with_cache().await?;
+        let _: Vec<Option<String>> = redis::cmd("MGET")
+            .arg("key_1")
+            .arg("key_2")
+            .query_async(&mut con)
+            .await?;
+        {
+            let s = con.get_cache_statistics();
+            assert_eq!(s.hit, 0);
+            assert_eq!(s.miss, 2);
+        }
+
+        let _: Vec<Option<String>> = redis::cmd("MGET")
+            .arg("key_1")
+            .arg("key_2")
+            .query_async(&mut con)
+            .await?;
+        {
+            let s = con.get_cache_statistics();
+            assert_eq!(s.hit, 2);
+            assert_eq!(s.miss, 2);
+        }
+        let _: Option<String> = redis::cmd("GET").arg("key_1").query_async(&mut con).await?;
+        {
+            let s = con.get_cache_statistics();
+            assert_eq!(s.hit, 3);
+            assert_eq!(s.miss, 2);
+        }
+        Ok::<_, RedisError>(())
+    })
+    .unwrap();
+}
+
+#[test]
+fn test_cache_drops_and_guards() {
+    use redis::ProtocolVersion;
+
+    let ctx = TestContext::new();
+    if ctx.protocol == ProtocolVersion::RESP2 {
+        return;
+    }
+
+    block_on_all(async move {
+        let con = ctx
+            .multiplexed_async_connection_tokio_with_cache_config(CacheConfig::default())
+            .await?;
+
+        let mut v = vec![];
+        for i in 0..1000 {
+            let mut c = con.clone();
+            v.push(async move {
+                let x: RedisFuture<Option<u32>> = c.get("KEY");
+                tokio::time::timeout(Duration::from_micros(i), x).await
+            })
+        }
+        futures::future::join_all(v).await;
+        let s = con.get_cache_statistics();
+        assert_eq!(s.miss, 1000);
+        Ok::<_, RedisError>(())
+    })
+    .unwrap();
+}
+
+#[test]
+fn test_cache_guards_and_different_types() {
+    use redis::ProtocolVersion;
+
+    let ctx = TestContext::new();
+    if ctx.protocol == ProtocolVersion::RESP2 {
+        return;
+    }
+
+    block_on_all(async move {
+        let mut con = ctx
+            .multiplexed_async_connection_tokio_with_cache_config(CacheConfig::default())
+            .await?;
+        let _: () = con.set("KEY", "77").await?;
+        let mut v = vec![];
+        for _ in 0..5 {
+            let mut c = con.clone();
+            v.push(tokio::spawn(async move {
```

Comment:
why are the spawns needed?

Reply:
For more reliability in tests, to make sure it doesn't rely in order of commands. Can be removed but I see no harm.

Reply:
this is fine - just add a comment explaining this, please.